### ISSUE TITLE
[DTensor] `to_local` + `from_local` on `Partial` DTensor produce different gradient

### ISSUE THREAD

**Issue by KimbingNg** [pytorch/pytorch#172932]
## Description

When using `DTensor.to_local()` followed by `DTensor.from_local()` on a DTensor with `Partial` placement, the forward pass appears to be a no-op (as expected), but the backward pass produces different gradients compared to using the DTensor directly.

Why aren't the gradients identical when using `to_local()` followed by `from_local()` with the same placements? Shouldn't this pattern be a no-op, resulting in the same gradients as using the DTensor directly?

This design is error-prone in some scenarios. Would it be better to add an extra check in `DTensor.to_local` to raise an error if the placements is `Partial` and `grad_placements` is not `Replicate()`?

## Reproduce Code
```py
# torchrun --nproc_per_node=2 test_grad.py
import torch
import loguru
from torch.distributed import init_process_group
from torch.distributed.device_mesh import init_device_mesh
from torch.distributed._tensor import DTensor, distribute_tensor, Shard

init_process_group()
torch.cuda.set_device(torch.distributed.get_rank())
device_mesh = init_device_mesh(device_type='cuda', mesh_shape=(2,), mesh_dim_names=('tp',))

def test_case(plan=1):
    original_tensor = torch.tensor([1., 2.], device='cuda', requires_grad=True)
    original_tensor = distribute_tensor(original_tensor, device_mesh, [Shard(0)])
    out = original_tensor.sum()
    if plan == 1:
        out = DTensor.from_local(
            out.to_local(), out.device_mesh, out.placements, run_check=False
        )
    elif plan == 2:
        pass # do nothing

    loss = out.full_tensor().sum()
    loss.backward()
    grad = original_tensor.grad.full_tensor()
    return loss, grad

loss1, grad1 = test_case(1)
loss2, grad2 = test_case(2)
loguru.logger.info(f"Plan 1 Loss: {loss1} Grad: {grad1}")
loguru.logger.info(f"Plan 2 Loss: {loss2} Grad: {grad2}")
```

## Output
```
Plan 1 Loss: 3.0 Grad: tensor([2., 2.], device='cuda:1')
Plan 1 Loss: 3.0 Grad: tensor([2., 2.], device='cuda:0')
Plan 2 Loss: 3.0 Grad: tensor([1., 1.], device='cuda:1')
Plan 2 Loss: 3.0 Grad: tensor([1., 1.], device='cuda:0')
```

### Versions

2.9.1

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @tianyu-l @XilunWu @SherlockNoMad @ppwwyyxx
